### ISSUE TITLE
dask.dataframe groupby(as_index=False) does nothing

### ISSUE THREAD

**Issue by thrasibule** [blaze/dask#735]
When doing a groupby, the index gets set even when specifying as_index=False. This combined with issue #734 means that groupby has to be the last operation, otherwise we're losing columns.
Also it seems that groupby works fine with multiple columns, even though multiindex are not supported, which doesn't seem very consistent.

**Comment by mrocklin** [comment]
Yes, we should at least raise a `NotImplementedError` here.

Can you give an example of losing columns?

**Comment by thrasibule** [comment]
What I mean by losing columns is that once a column becomes the index in dask, I see no way to bring it back as a normal column. If I use set_index(drop=False) that would work, but groupby silently creates an index without giving that option. The example is a little contrived, but hopefully conveys what I'm trying to achieve:

```python
df = pd.DataFrame({'A': [1,1,2,2,3,3], 'B': [1,2,-1, 2, 3, 4]})
df2= df.groupby('A', as_index=False).max()
df3 = df.groupby('B').mean()
```
In dask, A would be the index of df2, and then I don't see any way to operate on A, which is what I mean by losing columns.

**Comment by mrocklin** [comment]
OK, this makes sense to me.  I'm a bit saturated this week but may be able to pick this up next week.  At the very least `reset_index` should be trivial to add if we don't mind having a non-sensical index (e.g. `[0, 1, 2, 3, ... 0, 1, 2, 3, ... 0, 1, 2, 3]`)

**Comment by mrocklin** [comment]
Of course it'd be great if anyone else can pick this up in the meantime.

**Comment by cowlicks** [comment]
I can pick this up.

**Comment by cowlicks** [comment]
I'm confused by this. What was the approach you had in mind @mrocklin ?

**Comment by mrocklin** [comment]
We should probably check keyword arguments in groupby and raise notimplementederrors accordingly.  We should probably also implement `reset_index` which will just map `self._partition_type.reset_index` onto the blocks and set `divisions = [None] * (self.npartitions + 1)`

**Comment by cowlicks** [comment]
I don't think I've yet fixed @thrasibule's problem. Wouldn't we have to allow `as_index=False` I don't know the group by code very well, would this be too hard?

**Comment by mrocklin** [comment]
is `groupby(..., as_index=False)...` equivalent to `groupby(...)....reset_index()`?

**Comment by cowlicks** [comment]
Yes. But the current implementation drops the name of the index, and just calls it `'index'`. Which is what pandas does when the index is not named. Fixing this now.

**Comment by cowlicks** [comment]
This is kind of tricky to add since we are not tracking the index name like we do columns. I'm still figuring out how to do this.

**Comment by thebeancounter** [comment]
is this issue still open or the functionality is already available? 
please not this [stackoverflow question](https://stackoverflow.com/questions/47216564/python-dask-dataframes-concatenate-groupby-apply-output-to-a-single-data-frame) that could really use some clarification in the manner.

**Comment by thebeancounter** [comment]
@cowlicks

**Comment by kingfischer16** [comment]
This doesn't appear to be fixed (using dask version 1.1.5, Python 3.7.0). Performing a simple `.groupby` raises the following:
`TypeError: __init__() got an unexpected keyword argument 'as_index'`
In dask\dataframe\core.py the DataFrameGroupBy class is constructed and **kwargs are passed. The DataFrameGroupBy class uses the _GroupBy class from, dask\dataframe\groupby.py, as it's base (and I'm assuming uses its constructor as well), however the constructor of _GroupBy is defined as:
```
class _GroupBy(object):
    def __init__(self, df, by=None, slice=None):
```
I can see that _GroupBy inherits from object class, but I don't think that helps with accepting kwargs. 

If I add **kwargs to the constructor and the meta.groupby I no longer get the error...
```
[764] def __init__(self, df, by=None, slice=None, **kwargs):
         ...
[797]    self._meta = self.obj._meta.groupby(index_meta, **kwargs)
```
But the result is exactly the same as if as_index is set to True (default). Can this issue be reopened, or could somebody direct me to the current issue that has absorbed this? I still suspect that this can be fixed by passing the kwargs down through to the right level, but I have yet to get past the obj._meta.groupby() statement.

**Comment by mrocklin** [comment]
Thanks @thrasibule .  At this point I recommend raising a new issue with a minimal reproducible example, if you have the time.
